docs/proposals/dynamic-kubelet-configuration.md

```diff
@@ -0,0 +1,171 @@
+- The Kubelet's configuration type will no longer align with it's flags; we should add a separate struct that contains the flag variables to prevent them from migrating all over the codebase (similar to work being done in [#32215](https://github.com/kubernetes/kubernetes/issues/32215)).
+- We should add more structure to the Kubelet configuration for readability, and I would like to move the location of the type closer to the Kubelet in the K8s tree. The details of this portion can be discussed via a separate refactoring PR (**TODO** add a link here once I open that PR).
+- We need to be able to add and remove experimental fields from the `KubeletConfiguration` without having to rev the API version. A simple solution is to just have a string representation of the experimental fields as part of the KubeletConfiguration, that the Kubelet can parse as necessary. 
+    + Some additional thought needs to go toward the graduation policy for experimental fields. Will the kubelet continue to try to parse them out of the experimental section? For how long? What if the first-class field and experimental field conflict?
+
+
+### Representation and Organization of Kubelet Configuration in a Cluster
+
+- As a cluster-level object, Kubelet's configuration should be stored in a `ConfigMap` object.
+- On local disk, the Kubelet's configuration should be stored in a `.json` or `.yaml` file.
+- The Kubelet's configuration should be, at least initially, organized in the cluster as a structured monolith. 
+    + *Structured*, so that it is readable.
+    + *Monolithic*, to provide atomicity over the entire configuration object.
+        * This likely means that the Kubelet configuration will be stored as a string blob (JSON or YAML) in the value associated with a given key on a `ConfigMap`.
+        * If leaky boundaries occur between the substructures, we don't want the problem of coordinating non-atomic updates across separately-referenced substructures.
+        * If, in the future, the ability to independently orchestrate the substructures of the configuration is desired, we can move down that road. But today this is probably overkill, because most K8s cluster configuration is eventually homogeneous anyway. Even in a non-homogeneously configured cluster, we would have to carefully consider the downside of losing atomicity on the config object against the upside of more flexibility for splitting up configuration responsibility and e.g. being able to roll out changes to separate subcomponents at different rates.
+
+An example of what the `ConfigMap` object for just the Kubelet could look like:
+```
+kind: ConfigMap
+metadata:
+  name: my-kubelet-config-<hash of `data` for verification>
+data:
+ config: "{JSON representation of config, for example}"
+```
+
+If this idea extended so that all `Node` config is packaged together:
```

Comment:
This started as a Kubelet-specific idea, and it's still Kubelet-scoped for this quarter, but the direction it's going is probably a Node-specific ConfigMap. It's useful to be able to bundle node infra component config for a rollout, so that you don't have to worry about config skew between components. And you're right that this requires something special to plumb the config the Kubelet pulls down into the Pods it is managing. One idea was to have a special volume source for Node config: https://github.com/kubernetes/kubernetes/pull/29459#issuecomment-276163455